// yoco/lib/index.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.composable = exports.map = undefined;

var _reduxFlo = require('redux-flo');

var _reduxFlo2 = _interopRequireDefault(_reduxFlo);

var _bindMiddleware = require('@f/bind-middleware');

var _bindMiddleware2 = _interopRequireDefault(_bindMiddleware);

var _isArray = require('@f/is-array');

var _isArray2 = _interopRequireDefault(_isArray);

var _toMiddleware = require('@f/to-middleware');

var _toMiddleware2 = _interopRequireDefault(_toMiddleware);

var _mapArray = require('@f/map-array');

var _mapArray2 = _interopRequireDefault(_mapArray);

var _logError = require('@f/log-error');

var _logError2 = _interopRequireDefault(_logError);

var _channel = require('@f/channel');

var _channel2 = _interopRequireDefault(_channel);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _toConsumableArray(arr) { if (Array.isArray(arr)) { for (var i = 0, arr2 = Array(arr.length); i < arr.length; i++) { arr2[i] = arr[i]; } return arr2; } else { return Array.from(arr); } } /**
                                                                                                                                                                                                     * Imports
                                                                                                                                                                                                     */

/**
 * yoco
 * @param  {Array} mw redux style middleware
 * @return {Function} dispatch function to middleware
 */

function yoco(mw) {
  if (!(0, _isArray2.default)(mw)) mw = [mw];
  var dispatch = (0, _bindMiddleware2.default)([(0, _reduxFlo2.default)(errorHandler)].concat(_toConsumableArray(mw)));
  dispatch.wrap = wrap;
  dispatch.onError = _logError2.default;
  return dispatch;

  function errorHandler(err) {
    return dispatch.onError(err);
  }

  function wrap(fn) {
    return function () {
      var args = new Array(arguments.length);
      for (var i = 0; i < arguments.length; ++i) {
        args[i] = arguments[i];
      }
      return dispatch(fn.apply(this, args));
    };
  }
}

/**
 * crate a co from functions
 * @param {Array} fns mapping functions
 * @returm {Function}
 */

function map(fns) {
  if (!(0, _isArray2.default)(fns)) fns = [fns];
  return yoco((0, _mapArray2.default)(_toMiddleware2.default, fns));
}

var composable = function composable(mw) {
  return function (proc) {
    return regeneratorRuntime.mark(function _callee() {
      var _run,
          IN,
          OUT,
          isError,
          done,
          res,
          _args = arguments;

      return regeneratorRuntime.wrap(function _callee$(_context) {
        while (1) {
          switch (_context.prev = _context.next) {
            case 0:
              _run = run(proc.apply(undefined, _args));
              IN = _run.IN;
              OUT = _run.OUT;
              isError = _run.isError;
              done = _run.done;

            case 5:
              if (!true) {
                _context.next = 24;
                break;
              }

              _context.next = 8;
              return OUT.take();

            case 8:
              res = _context.sent;

              if (!isError()) {
                _context.next = 13;
                break;
              }

              throw res;

            case 13:
              if (done()) {
                _context.next = 21;
                break;
              }

              _context.t0 = IN;
              _context.next = 17;
              return res;

            case 17:
              _context.t1 = _context.sent;

              _context.t0.put.call(_context.t0, _context.t1);

              _context.next = 22;
              break;

            case 21:
              return _context.abrupt('return', res);

            case 22:
              _context.next = 5;
              break;

            case 24:
            case 'end':
              return _context.stop();
          }
        }
      }, _callee, this);
    });

    function run(action) {
      var OUT = (0, _channel2.default)();
      var IN = (0, _channel2.default)();
      var _done = false;
      var _isError = false;
      var dispatch = yoco([].concat(_toConsumableArray(mw), [function (ctx) {
        return function (next) {
          return function (action) {
            OUT.put(action);
            return IN.take();
          };
        };
      }]));
      dispatch.onError = function (err) {
        _isError = true;
        OUT.put(err);
      };
      dispatch(action).then(function (val) {
        _done = true;
        OUT.put(val);
      });
      function done() {
        return _done;
      }
      function isError() {
        return _isError;
      }
      return { IN: IN, OUT: OUT, done: done, isError: isError };
    }
  };
};

/**
 * Exports
 */

exports.default = yoco;
exports.map = map;
exports.composable = composable;
